packages/core/src/app/pg-core/db/chain-db.ts

```diff
@@ -1,24 +1,523 @@
+/* External Imports */
+import BigNum = require('bn.js')
+
 /* Internal Imports */
-import { KeyValueStore, ChainDB } from '../../../interfaces'
-import { BaseKey } from '../../common'
+import {
+  KeyValueStore,
+  ChainDB,
+  Transaction,
+  InclusionProof,
+} from '../../../interfaces'
+import {
+  BaseKey,
+  encode,
+  decode,
+  bnToUint256,
+  getTransactionRangeEnd,
+} from '../../common'
+
+// TODO: Where should this stuff sit? Should it be customizeable?
+const computeParent = (leftChild: string, rightChild: string): string => {
+  return
+}
 
 /**
- * Basic ChainDB implementation that provides a
- * nice interface to the chain database.
+ * Checks whether a node is a left child.
+ * @param nodeIndex Node index to check
+ * @returns `true` if the node is a left child, `false` otherwise.
  */
-export class PGChainDB implements ChainDB {
-  private statedb: KeyValueStore
-  private historydb: KeyValueStore
+const isLeftChild = (nodeIndex: number): boolean => {
```

Comment:
@karlfloersch Do you think it makes sense for these binary tree utilities to live in a separate utils file? Might make them easier to test.

Reply:
Hmm I've gone back and forth about this. The ChainDB is quite large and so it would be nice to break these out into another file. Plus they could be useful for other implementations of ChainDB that use merkle trees. That said, if it's in another file I do feel that file should live very close to this ChainDB file so it doesn't just get added to a massive utils file with random stuff in it.